### PR TITLE
feat: [experimental] enable sum loss trainer

### PR DESCRIPTION
**Pull request by dushyantbehl**
### Description of the change

Try to enable loss reduction to sum with hf trainer.

### Was the PR tested

- [x] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

File: tuning/config/configs.py [hunk 1]
```diff
@@ -208,6 +208,17 @@ class TrainingArguments(transformers.TrainingArguments):
...
     )
+    enable_reduce_loss_sum: bool = field(
+        default=False,
+        metadata={
+            "help": "Pass `True` to enable use of sum loss reduction on the loss function. \
+                Switches out the compute_loss in SFTTrainer with that from internal one in\
+                tuning.trainers.sum_loss_sft_trainer::SumLossSFTTrainer. \
+                Please note this feature is experimental and not fully supported. \
+                One Known limitation of this function is PEFT PT so its disabled \
+                for all PEFT runs by the library internally."
```
Comment (ashokponkumar) [inline tuning/config/configs.py:219]:
```suggestion
            "help": "Pass `True` to enable use of sum loss reduction on the loss function. \
                Please note this feature is experimental and not fully supported. \
                PEFT is not supported with this flag yet."
```

File: tuning/sft_trainer.py [hunk 2]
```diff
@@ -339,16 +341,40 @@ def train(
...
-        train_dataset=formatted_train_dataset,
-        eval_dataset=formatted_validation_dataset,
-        data_collator=data_collator,
-        args=training_args,
-        callbacks=trainer_callbacks,
-        peft_config=peft_config,
-    )
+    trainer = None
+    if train_args.enable_reduce_loss_sum:
+        trainer = SumLossSFTTrainer(
```
Comment (ashokponkumar) [inline tuning/sft_trainer.py:346]: If the parameters are same across both, would it make sense to have a variable to capture the class to use, and do the initialisation call only once?

File: tuning/sft_trainer.py [hunk 3]
```diff
@@ -339,16 +341,40 @@ def train(
...
+            eval_dataset=formatted_validation_dataset,
+            data_collator=data_collator,
+            args=training_args,
+            callbacks=trainer_callbacks,
+            peft_config=peft_config,
+        )
+
+        # Note this check has to be done post trainer initialization
+        if _is_peft_model(trainer.model):
+            logger.warning(
```
Comment (ashokponkumar) [inline tuning/sft_trainer.py:359]: Shouldn't we rather fail if both enable_reduce_loss_sum and peft flags are enabled?

**Review by kmehant — COMMENTED** [review]
I am skeptical in introducing a new class, 

I wanted to make sure if we have seen these two flows

Causal Loss flow in transformers

1. https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/loss/loss_utils.py#L41
2. https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/loss/loss_utils.py#L27-L38 (loss reduction is chosen based on num_items_in_batch)

Custom loss over outputs and labels after forward pass through `compute_loss_func`. You might want to see the flow of self.compute_loss_func in https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/trainer.py#L3778 (loss reduction can be applied in a custom func and passed to existing trainer)

Let me know.

**Comment by dushyantbehl** [comment]
> I am skeptical in introducing a new class,

@kmehant what is the skepticism behind it can you elaborate

> I wanted to make sure if we have seen these two flows
> 
> Causal Loss flow in transformers
> 
> 1. https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/loss/loss_utils.py#L41
> 2. https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/loss/loss_utils.py#L27-L38

I am aware of the trainer loss flow and if you see [this](https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/loss/loss_utils.py#L27-L38) the function alread performs an averaging over `num_items_in_batch`

> Custom loss over outputs and labels after forward pass through `compute_loss_func`. You might want to see the flow of self.compute_loss_func in https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/trainer.py#L3778
> 

Also custom `compute_loss_func` doesn't work here because when computing full loss outside requires you to use state like `model embedding` size and was not suitable for when I tried it. 

Maybe we could revisit the flow at later point in time but did not seem like a HARD problem to me because [HF themselves call out ](https://github.com/huggingface/transformers/blob/08e3217bafddc5d11ce0e7369bcfaaabe5501ba5/src/transformers/trainer.py#L3782)that you should

`Subclass and override for custom behavior`

File: tuning/sft_trainer.py (hunk 3 above)
Comment (dushyantbehl) [reply tuning/sft_trainer.py:359]: its up to us to decide weather to fail or continue with a user message..if we decide to fail the above flow makes more sense to me.

https://github.com/foundation-model-stack/fms-hf-tuning/pull/520#discussion_r2034832245

File: tuning/sft_trainer.py (hunk 2 above)
Comment (dushyantbehl) [reply tuning/sft_trainer.py:346]: you would have to initialize a class anyway to know if the model is PEFT or not 
https://github.com/foundation-model-stack/fms-hf-tuning/pull/520/files#diff-7d8f6f8626340337d03a2f04fec2840d2ede32b0c7af2640eed9759af45bb3cfR357

File: tuning/sft_trainer.py [hunk 4]
```diff
@@ -350,6 +357,13 @@ def train(
         peft_config=peft_config,
     )
 
+    # Note this check has to be done post trainer initialization
+    if train_args.enable_reduce_loss_sum and _is_peft_model(trainer.model):
+        raise ValueError(
+            "❌ Loaded model is PEFT model but both PEFT and sum loss should not be set together."
```
Comment (ashokponkumar) [inline tuning/sft_trainer.py:363]:
```suggestion
            "❌ Loaded model is PEFT model but both PEFT and sum loss is not supported yet."
```

**Review by ashokponkumar — DISMISSED** [review]
